Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -7,319 +7,354 @@
+                                break;
+                        }
+                    }
+                }
 
-		public override object Clone()
-		{
-			var clone = new ChainHeadersBehavior(this.Chain, this.SharedState)
-			{
-				CanSync = this.CanSync,
-				CanRespondToGetHeaders = this.CanRespondToGetHeaders,
-				AutoSync = this.AutoSync,
-				_State = this._State
-			};
-			return clone;
-		}
+                this.AttachedNode.SendMessageAsync(headers);
+            }
+
+            // == HeadersPayload ==
+            // Represents the peers height from our point view.
+            // This updates the pending tip parameter which is 
+            // the peers current best validated height.
+            // If the peer's height is higher Chain.Tip is updated to have 
+            // the most PoW header.
+            // It is sent in response to GetHeadersPayload or is solicited by the 
+            // peer when a new block is validated (and not in IBD).
+
+            var newHeaders = message.Message.Payload as HeadersPayload;
+            if ((newHeaders != null) && this.CanSync)
+            {
+                ChainedBlock pendingTipBefore = this.GetPendingTipOrChainTip();
+                this.logger.LogTrace("Pending tip is '{0}', received {1} new headers.", pendingTipBefore, newHeaders.Headers.Count);
+
+                // TODO: implement MAX_HEADERS_RESULTS in NBitcoin.HeadersPayload
+
+                ChainedBlock tip = pendingTipBefore;
+                foreach (BlockHeader header in newHeaders.Headers)
+                {
+                    ChainedBlock prev = tip.FindAncestorOrSelf(header.HashPrevBlock);
+                    if (prev == null)
+                        break;
+
+                    tip = new ChainedBlock(header, header.GetHash(), prev);
+                    bool validated = this.Chain.GetBlock(tip.HashBlock) != null || tip.Validate(this.AttachedNode.Network);
+                    validated &= !this.chainState.IsMarkedInvalid(tip.HashBlock);
+                    if (!validated)
+                    {
+                        this.logger.LogTrace("Validation of new header '{0}' failed.", tip);
+                        this.invalidHeaderReceived = true;
+                        break;
+                    }
+
+                    this.pendingTip = tip;
+                }
+
+                if (pendingTipBefore != this.pendingTip)
+                    this.logger.LogTrace("Pending tip changed to '{0}'.", this.pendingTip);
+
+                // Long reorganization protection on POS networks.
+                bool reorgPrevented = false;
+                uint maxReorgLength = this.chainState.MaxReorgLength;
+                Network network = this.AttachedNode?.Network;
+                ChainedBlock consensusTip = this.chainState.ConsensusTip;
+                if ((maxReorgLength != 0) && (network != null) && (consensusTip != null))
```

Comment:
Logically this looks fine, semantically I would suggest to put the whole reorg bloc inside its own `if` statement and avoid the allocations of network and consensusTip
It make the code more readable. 

```
 bool reorgPrevented = false;
  if(this.preventReorgs)
  {
                uint maxReorgLength = this.chainState.MaxReorgLength;
                Network network = this.AttachedNode?.Network;
                ChainedBlock consensusTip = this.chainState.ConsensusTip;
                if ((network != null) && (consensusTip != null))
                {
                         bla..
                }
  }
``` 
You could even have a flag at the constructor that specifies if prevention is enabled
```
this.preventReorgs = this.chainState.MaxReorgLength != 0
```

Reply:
I see, ok, will fix

Reply:
I can't do the constructor thing, MaxReorgLength is not initialized at that point, but except for that, i'll do as you suggest